flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,351 @@
+			this.kubeClient.logException(e);
+			return false;
+		}
+		workerNodeMap.remove(worker.getResourceID());
+		return true;
+	}
+
+	@Override
+	public void onAdded(List<FlinkPod> pods) {
+		for (FlinkPod pod : pods) {
+			log.info("Received new TaskExecutor pod: {} - Remaining pending pod requests: {}",
+				pod.getName(), numPendingPodRequests);
+
+			if (numPendingPodRequests > 0) {
+				numPendingPodRequests--;
+				KubernetesWorkerNode worker = new KubernetesWorkerNode(new ResourceID(pod.getName()));
+				workerNodeMap.put(worker.getResourceID(), worker);
+			} else {
+				log.info("Returning excess pod {}.", pod.getName());
+				kubeClient.stopPod(pod.getName());
+			}
+		}
+	}
+
+	@Override
+	public void onModified(List<FlinkPod> pods) {
+		pods.forEach(this::removePodIfTerminated);
+	}
+
+	@Override
+	public void onDeleted(List<FlinkPod> pods) {
+		pods.forEach(this::removePodIfTerminated);
+	}
+
+	@Override
+	public void onError(List<FlinkPod> pods) {
+		pods.forEach(this::removePodIfTerminated);
+	}
+
+	protected ConcurrentMap<ResourceID, KubernetesWorkerNode> getWorkerNodeMap() {
+		return workerNodeMap;
+	}
+
+	private void requestKubernetesPod() {
+		Preconditions.checkNotNull(this.kubeClient);
+
+		numPendingPodRequests++;
+
+		log.info("Requesting new TaskExecutor pod with <{},{}>. Number pending requests {}.",
+			defaultTaskManagerMemoryMB,
+			defaultCpus,
+			numPendingPodRequests);
+
+		String podName = clusterId + Constants.NAME_SEPARATOR + TASK_MANAGER_PREFIX +
+			Constants.NAME_SEPARATOR + maxPodId.addAndGet(1);
+
+		try {
+			HashMap<String, String> env = new HashMap<>();
+			env.put(ENV_RESOURCE_ID, podName);
+
+			TaskManagerPodParameter parameter = new TaskManagerPodParameter(
+				podName,
+				getTaskManagerStartCommand(podName),
+				defaultTaskManagerMemoryMB,
+				defaultCpus,
+				env);
+
+			this.kubeClient.createTaskManagerPod(parameter);
+		} catch (Exception e) {
+			this.kubeClient.logException(e);
+			throw new FlinkRuntimeException("Could not start new worker");
+		}
+	}
+
+	/**
+	 * Request new pod if pending pods cannot satisfies pending slot requests.
+	 */
+	private void requestKubernetesPodIfRequired() {
+		int requiredTaskManagerSlots = getNumberRequiredTaskManagerSlots();
+		int pendingTaskManagerSlots = numPendingPodRequests * numberOfTaskSlots;
+
+		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+			requestKubernetesPod();
+		}
+	}
+
+	private void removePodIfTerminated(FlinkPod pod) {
+		if (pod.isTerminated()) {
+			kubeClient.stopPod(pod.getName());
+			final KubernetesWorkerNode kubernetesWorkerNode = workerNodeMap.remove(new ResourceID(pod.getName()));
+			if (kubernetesWorkerNode != null) {
+				requestKubernetesPodIfRequired();
+			}
+		}
+	}
+
+	private List<String> getTaskManagerStartCommand(String podName) {
+		final ContaineredTaskManagerParameters taskManagerParameters =
+			ContaineredTaskManagerParameters.create(flinkConfig, 1333, numberOfTaskSlots);
```

Comment:
Shall we use a parameter instead of the constant 1333 in ContaineredTaskManagerParameters.create()?

Reply:
Thansk @thousandhu. It is bug. I will fix it.